### PR TITLE
Fix bor txlookup

### PR DESCRIPTION
**Pull request by enriavil1**
Currently we are not saving bor state-sync transactions into our `TxLookUpEntry`

### CODE REVIEW

**Review by AskAlexSharov — CHANGES_REQUESTED** [review]
we have `eth/stagedsync/stage_txlookup.go` to write TxLookup. Can we write it there and in sorted way?

**Comment by 0xKrishna** [comment]
Just for reference : https://github.com/maticnetwork/erigon/commit/fe37ff920a790fc6126ba8718d3f0858da179a8e
This will save some changes.

**Comment by 0xKrishna** [comment]
I am still using v2022.07.04 and the above changes are on top of that, So some params might differ.

**Comment by 0xKrishna** [comment]
Looking better, Can we remove some unwanted changes now, I see some changes in this PR which are not required now.

File: cmd/rpcdaemon/commands/eth_receipts.go [hunk 1]
```diff
@@ -275,7 +275,7 @@ func (api *APIImpl) GetTransactionReceipt(ctx context.Context, txnHash common.Ha
 	var ok bool
 
 	blockNum, ok, err = api.txnLookup(ctx, tx, txnHash)
-	if !ok || blockNum == 0 {
+	if !ok {
```
Comment (0xKrishna) [inline cmd/rpcdaemon/commands/eth_receipts.go:278]: Is this change needed now ?

File: eth/stagedsync/stage_txlookup.go [hunk 2]
```diff
@@ -116,10 +120,10 @@ func txnLookupTransform(logPrefix string, tx kv.RwTx, blockFrom, blockTo uint64,
 				return err
 			}
 		}
-
-		if cfg.isBor {
-			borPrefix := []byte("matic-bor-receipt-")
-			if err := next(k, crypto.Keccak256(append(append(borPrefix, k...), v...)), blockNumBytes); err != nil {
+		// we add state sync transactions every bor Sprint amount of blocks
+		if cfg.isBor && blocknum%cfg.borSprint == 0 {
```
Comment (0xKrishna) [inline eth/stagedsync/stage_txlookup.go:124]: Change it to 
```
if cfg.isBor && blocknum%cfg.borSprint == 0 && rawdb.HasBorReceipts(tx, blocknum) {
```

Because it's not necessary that every 64 block there will be state-sync, I mean it's possible there is one at blcok 64 and none at 128 and then again one at 192.

File: cmd/rpcdaemon/commands/eth_receipts.go (hunk 1 above)
Comment (enriavil1) [reply cmd/rpcdaemon/commands/eth_receipts.go:278]: No but it does not affect us.

File: turbo/snapshotsync/block_snapshots.go [hunk 3]
```diff
@@ -1512,7 +1538,7 @@ func expectedTxsAmount(snapDir string, blockFrom, blockTo uint64) (firstTxID, ex
 	return
 }
 
-func TransactionsIdx(ctx context.Context, chainID uint256.Int, blockFrom, blockTo uint64, snapDir string, tmpDir string, p *background.Progress, lvl log.Lvl) (err error) {
+func TransactionsIdx(ctx context.Context, chainID uint256.Int, tx kv.Tx, blockFrom, blockTo uint64, snapDir string, tmpDir string, p *background.Progress, lvl log.Lvl, isBor bool, sprint uint64) (err error) {
```
Comment (AskAlexSharov) [inline turbo/snapshotsync/block_snapshots.go:1541]: let's combine Bor params to 1 variable?
something like: *BorConfig
or something like: struct{ Enabled bool, Sprint uint64)

File: turbo/snapshotsync/block_snapshots.go [hunk 4]
```diff
@@ -1621,19 +1653,38 @@ RETRY:
...
 				}
 			}
 
-			if err := txnHashIdx.AddKey(slot.IDHash[:], offset); err != nil {
-				return err
-			}
-			if err := txnHash2BlockNumIdx.AddKey(slot.IDHash[:], blockNum); err != nil {
-				return err
+			if isBorTx {
+				header := rawdb.ReadHeaderByNumber(tx, blockNum)
```
Comment (AskAlexSharov) [inline turbo/snapshotsync/block_snapshots.go:1670]: - to get header hash can use `ReadCanonicalHash`
- but Snapshots Indexing happening after downloading of snapshots - db is empty in that time.

**Review by AskAlexSharov — CHANGES_REQUESTED** [review]

File: turbo/snapshotsync/block_snapshots.go (hunk 4 above)
Comment (AskAlexSharov) [reply turbo/snapshotsync/block_snapshots.go:1670]: see how `bodyGetter` used instead of db in this method

File: turbo/snapshotsync/block_snapshots.go [hunk 5]
```diff
@@ -1622,32 +1670,47 @@ RETRY:
...
 				}
+				headerBuf, _ = headerGetter.Next(headerBuf[:0])
+				if err := rlp.DecodeBytes(headerBuf, header); err != nil {
+					return err
+				}
+
 				blockNum++
 			}
-
+			txTypeAndlengthOfAddress := 21
```
Comment (AskAlexSharov) [inline turbo/snapshotsync/block_snapshots.go:1685]: 1 it's not txType - it's first byte for txnHash (to reduce false-positives produced by our indices).

File: turbo/snapshotsync/block_snapshots.go [hunk 6]
```diff
@@ -757,14 +757,14 @@ func buildIdx(ctx context.Context, sn snap.FileInfo, chainID uint256.Int, tmpDir
...
-		if err := TransactionsIdx(ctx, chainID, sn.From, sn.To, dir, tmpDir, p, lvl); err != nil {
+		if err := TransactionsIdx(ctx, chainID, tx, sn.From, sn.To, dir, tmpDir, p, lvl, borCfg); err != nil {
 			return err
 		}
 	}
 	return nil
 }
 
-func BuildMissedIndices(ctx context.Context, dir string, chainID uint256.Int, tmpDir string, workers int, lvl log.Lvl) error {
+func BuildMissedIndices(ctx context.Context, dir string, chainID uint256.Int, db kv.Tx, tmpDir string, workers int, lvl log.Lvl, borCfg types.BorConfigSprint) error {
```
Comment (AskAlexSharov) [inline turbo/snapshotsync/block_snapshots.go:767]: need remove `db` parameter?

**Review by AskAlexSharov — CHANGES_REQUESTED** [review]

File: turbo/snapshotsync/block_snapshots.go [hunk 7]
```diff
@@ -1622,32 +1670,47 @@ RETRY:
...
 				}
 			}
-
-			if err := txnHashIdx.AddKey(slot.IDHash[:], offset); err != nil {
-				return err
-			}
-			if err := txnHash2BlockNumIdx.AddKey(slot.IDHash[:], blockNum); err != nil {
-				return err
+			emptySender := make([]byte, 20)
+			if slot.IsBor && bytes.Equal(word[2:txTypeAndlengthOfAddress], emptySender) {
```
[the remark this answers is not in the document]
Comment (AskAlexSharov) [reply turbo/snapshotsync/block_snapshots.go:1695]: `slot.IsBor` - can it be `true` on mainnet? If yes, then feels weird

**Review by AskAlexSharov — APPROVED** [review]
